### PR TITLE
Support %, %%, #, ## operators

### PR DESCRIPTION
**Pull request by pecigonzalo**
Add support for more interpolations

## Notes
Relates to #2

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1 +1,5 @@
 module github.com/buildkite/interpolate
+
+go 1.12
+
+require github.com/drone/envsubst v1.0.2
```
Comment (lox) [inline go.mod]: Did you mean to require this?

**Comment by lox** [comment]
Aside from the `github.com/drone/envsubst` require, this looks really good. Thanks for your time and patience @pecigonzalo!

Just checking, are all these new expansions standard bash ones?

File: go.mod (hunk 1 above)
Comment (pecigonzalo) [reply go.mod]: Yeah, its used by `substitutions.go` 

The official Go behaves a bit differently, here is the diff:

```
diff --git i/this.go w/this.go
index d39d244..9306b0c 100644
--- i/this.go
+++ w/this.go
@@ -6,14 +6,13 @@ package path
 
 import (
 	"errors"
-	"strings"
 	"unicode/utf8"
 )
 
-// ErrBadPattern indicates a pattern was malformed.
+// ErrBadPattern indicates a globbing pattern was malformed.
 var ErrBadPattern = errors.New("syntax error in pattern")
 
-// Match reports whether name matches the shell pattern.
+// Match reports whether name matches the shell file name pattern.
 // The pattern syntax is:
 //
 //	pattern:
@@ -43,7 +42,10 @@ Pattern:
 		star, chunk, pattern = scanChunk(pattern)
 		if star && chunk == "" {
 			// Trailing * matches rest of string unless it has a /.
-			return !strings.Contains(name, "/"), nil
+			// return !strings.Contains(name, "/"), nil
+
+			// Return rest of string
+			return true, nil
 		}
 		// Look for match at current position.
 		t, ok, err := matchChunk(chunk, name)
@@ -59,8 +61,7 @@ Pattern:
 		}
 		if star {
 			// Look for match skipping i+1 bytes.
-			// Cannot skip /.
-			for i := 0; i < len(name) && name[i] != '/'; i++ {
+			for i := 0; i < len(name); i++ {
 				t, ok, err := matchChunk(chunk, name[i+1:])
 				if ok {
 					// if we're the last chunk, make sure we exhausted the name
@@ -158,9 +159,6 @@ func matchChunk(chunk, s string) (rest string, ok bool, err error) {
 			}
 
 		case '?':
-			if s[0] == '/' {
-				return
-			}
 			_, n := utf8.DecodeRuneInString(s)
 			s = s[n:]
 			chunk = chunk[1:]
```

**Comment by pecigonzalo** [comment]
They are, check: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02 and https://www.tldp.org/LDP/abs/html/parameter-substitution.html

I was going to add `^^` and `^` but I was not so sure about those

**Comment by lox** [comment]
I'm uncomfortable bringing in `github.com/drone/envsubst/path`. They are an open-source competitor and it doesn't seem fair to them. 

Besides that, if we were going to bring in that dependency, we should just move entirely to that library and contribute upstream.

**Comment by pecigonzalo** [comment]
I understand, I would say that file is not the main thing tho, as its 3 lines modified on the file upstream, I would have to basically do the same and copy the file here. I thought it would be better to give them credit by referencing their project than re-implement it here.

I think it would be a great idea to "merge" the projects and just contribute to that one, I suggested using `drone/envsubst` directly over in Slack, but since that is a major change Im not sure how to PR that here, at least for the functions referenced here: https://github.com/buildkite/interpolate/pull/6/files#diff-b6d44dcddb0b7ad9e5118697cc8ddb3eR10 which I also tried to use from their upstream, but the package does not export them.

`drone/envsubst` its a bit more simple in the way it works (eg. does not have a `VariableExpansion` type/struct implementation) and many of their functions are not exported.

**Comment by pecigonzalo** [comment]
@lox what would be your preferred approach here?

**Comment by pecigonzalo** [comment]
Any updates on this?

**Comment by pecigonzalo** [comment]
Merged master to resolve conflicts.

**Comment by pecigonzalo** [comment]
@ticky @lox Any chance to merge this? It has been waiting for over 1yr :)

**Comment by lox** [comment]
I'm not too fussed about merging this in here, the complexity is that we'd need to match the same rules in Buildkite's backend parser. @yob thoughts?

**Pull request closed without merging**